### PR TITLE
add NFFT and pyNFFT

### PR DESCRIPTION
**Pull request by grlee77**
This PR adds the Non-equispaced fast Fourier transform (NFFT) library and its Python wrappers pyNFFT (similar to pyFFTW for FFTW).

A few notes:
- I have only implemented builds for linux and osx here.  It is possible to build NFFT for Windows, but this seems to require msys2 and MinGW-w64:  https://www-user.tu-chemnitz.de/~potts/nfft/faq.php

- gcc/libgcc on osx are for OpenMP support, which is assumed by the pyNFFT wrappers.

- pyNFFT only supports NFFT 3.2.x, but the most recent version of NFFT is 3.3.1.  Is there a way to make both 3.2.x and 3.3.x versions of NFFT available?  If not, I would prefer to stay at 3.2.4 for pyNFFT compatibility.

- The CUnit package is needed for `make check` with NFFT v3.3.x, but is not used by 3.2.4 so it is currently commented out in the build requirements for NFFT's meta.yaml.  Building CUnit for Windows is supported via `jam` (https://www.freetype.org/jam/), but I haven't attempted to implement that here.

### CODE REVIEW

**Comment by grlee77** [comment]
I think appveyor can be stopped because there are no windows packages to build here

**Comment by scopatz** [comment]
> I think appveyor can be stopped because there are no windows packages to build here

you should put "[skip appveyor]" at the end of your commit message then

File: recipes/cunit/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+build:
+  number: 0
+  skip: True  # [win]
+
+requirements:
+  build:
+    - libtool  # [unix]
+    - autoconf  # [unix]
+    - automake  # [unix]
```
Comment (scopatz) [inline recipes/cunit/meta.yaml]: no need to select unix here since windows is already skipped.

**Comment by grlee77** [comment]
okay,  I will try to remember that in the future

**Comment by grlee77** [comment]
Hmm... looks like 2 out of the 3 recipes here are working.  Circle and Travis have the same failure on import during the final testing of pyNFFT.  This did not occur on my local machine, so I am not sure what the cause is yet.

**Comment by grlee77** [comment]
It might also be desirable to set pyNFFT as `nomkl`.  Sometimes the MKL FFT routines can shadow those in FFTW (I have seen this in the past with pyFFTW if numpy w/ mkl is imported first).  I know the MKL FFT routines only support a subset of the functionality in FFTW, but I am not sure if this is potentially a problem for the way FFTW is used by NFFT.

**Comment by scopatz** [comment]
> Hmm... looks like 2 out of the 3 recipes here are working.

It is probably also a good idea to have a separate PR for each recipe.  Otherwise it wastes build time as issues get worked through

File: recipes/pynfft/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  url: https://pypi.io/packages/source/p/pyNFFT/pyNFFT-{{ version }}.tar.gz
+  md5: d2457a3f43839a8acce3295c60d5e692
+
+build:
+  number: 0
+  skip: True  # [win]
+
+requirements:
+  build:
+    - numpy >=1.6
```
Comment (ocefpaf) [inline recipes/pynfft/meta.yaml]: I believe there are compiled extensions in this package, right? If so, and if they depend on the numpy ABI you should change `numpy >=1.6` to `numpy x.x`. We only build `numpy >=1.9`, so the version limitation is already satisfied, but the `x.x` ensure that the right numpy, used to build, will be installed at run time by baking the numpy build string in the package.

File: recipes/pynfft/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    - python >=2.7
+    - setuptools
+    - nfft >=3.2,<3.3
+    - gcc
+    - cython
+  run:
+    - numpy >=1.6
```
Comment (ocefpaf) [inline recipes/pynfft/meta.yaml]: If https://github.com/conda-forge/staged-recipes/pull/583/files#r62758763 is true than you need to use `numpy x.x` here too.

File: recipes/nfft/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  md5: 8d9b1b708236297397ed1b70cace8b7b
+
+build:
+  number: 0
+  skip: True  # [win]
+
+requirements:
+  build:
+    - fftw
+    - gcc  # [osx]
```
Comment (ocefpaf) [inline recipes/nfft/meta.yaml]: Why only `OS X`? If you need OMP or Fortran the gcc package from the default channel is out current option for both Linux and OS X builds.

Comment (grlee77) [inline recipes/nfft/meta.yaml]: I didn't realize the default gcc on linux did not already support OpenMP

**Comment by grlee77** [comment]
> It is probably also a good idea to have a separate PR for each recipe. Otherwise it wastes build time as issues get worked through

I was hoping to get lucky on the first try! :wink:
I can submit cunit and NFFT as separate PRs and rebase this one to only contain pyNFFT.

File: recipes/nfft/meta.yaml (hunk 4 above)
Comment (ocefpaf) [inline recipes/nfft/meta.yaml]: It does. The problem is that it you would ship a broken package without `libgmop` and or you would need to request the end user to install that from her/his package manager and expect it to be compatible. So the current choice, to be on the safe side, is to use the default channel one and add `libgcc` are a run time dependency.

**Comment by ocefpaf** [comment]
@grlee77 even though I agree with @scopatz that you should send separate PRs these package depend on each other and you might need to tweak one to make the other work. In this case it is OK to keep them in one PR.

(I am trying to debug the last failure here.)

File: recipes/cunit/build.sh [hunk 5]
```diff
@@ -0,0 +1,6 @@
+
```
Comment (ocefpaf) [inline recipes/cunit/build.sh]: Can you add shebangs to all `build.sh` please? I guess that conda does not mind, but it looks nicer with them :smile:

**Comment by grlee77** [comment]
@ocefpaf
I will keep them together, then.  cunit is not required here any more now that I rolled NFFT back to 3.2.4, but I figured we may as well keep the recipe as it may find use elsewhere.

**Comment by grlee77** [comment]
I can replicate the same import failure locally when I use Python 2.7.  I was using Python 3.5 while testing.

**Comment by ocefpaf** [comment]
@grlee77 I was unable to compile `nfft` locally using the default channel `gcc`.

**Comment by grlee77** [comment]
On both of those pyNFFT CI failures, it seems Python 3.5 is being used during build but then 2.7 during testing!  any idea what is causing that?  
e.g.





during build after "The following NEW packages will be INSTALLED"
```
    python:          3.5.1-0       defaults                                        
```
at the start of testing
```
TEST START: pynfft-1.3.2-0
Fetching package metadata: ........
Solving package specifications: .........

The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    libgcc-5.2.0               |                0         1.1 MB  defaults
    python-2.7.11              |                0        12.0 MB  defaults
```

**Comment by grlee77** [comment]
removing the >=2.7 version qualifiers from python in `meta.yaml` for pyNFFT fixed the above issue for me locally.  will push that change and see how it goes

**Comment by ocefpaf** [comment]
> removing the >=2.7 version qualifiers from python in `meta.yaml` for pyNFFT fixed the above issue for me locally.  will push that change and see how it goes

I was about to send a PR to your branch doing that :smile: 

Our CIs will create a build matrix for you with Python 2.7, 3.4, 3.5. Same for numpy when you use `numpy x.x`.

**Comment by grlee77** [comment]
I agree that the python versions in the build matrix are compatible so the >=2.7 was unecessary.   I would have expected it to work as it was, though.  Does the behaviour I encountered point to a bug in conda build or was I specifying the versions inappropriately?

**Comment by ocefpaf** [comment]
> I agree that the python versions in the build matrix are compatible so the >=2.7 was unecessary.   I would have expected it to work as it was, though.  Does the behaviour I encountered point to a bug in conda build or was I specifying the versions inappropriately?

I believe that call simple call to `conda-build` should work that way. The problem was that we use [call conda-build-all](https://github.com/conda-forge/staged-recipes/blob/master/.CI/build_all#L25), and that passes some extra options to `conda-build` to create the build matrix.

**Comment by grlee77** [comment]
I did also encountered the problem with mixed Python versions locally when calling conda-build as follows:
```
conda build --python 2.7 -c conda-forge pynfft
```
(The only difference in my local meta.yaml from the current PR is that I did not use `x.x` for numpy and I still had `python >=2.7` in both build and run)

**Comment by grlee77** [comment]
I guess the problem is probably that one should not specify a specific `--python` version on the command line while also having the version constraints in meta.yaml

**Comment by ocefpaf** [comment]
We have green!

File: recipes/nfft/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+{% set version = "3.2.4" %}
+
+package:
+  name: nfft
+  version: {{ version }}
+
+source:
+  fn: nfft-{{ version }}.tar.gz
+  # url: https://github.com/NFFT/nfft/releases/download/{{ version }}/nfft-{{ version }}.tar.gz
+  # md5: 8c0516e3252ff39a245e62f1a59d0165
```
Comment (ocefpaf) [inline recipes/nfft/meta.yaml]: Can you remove these commented lines for the final version?

File: recipes/nfft/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,49 @@
+#{% set version = "3.3.1" %}
```
Comment (ocefpaf) [inline recipes/nfft/meta.yaml]: I guess you can remove this too.

File: recipes/nfft/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+requirements:
+  build:
+    - fftw
+    - gcc
+    - libtool
+    - autoconf
+    - automake
+# CUnit needed for "make check" in pyNFFT 3.3.x, but not for 3.2.x
+#     - cunit
```
Comment (ocefpaf) [inline recipes/nfft/meta.yaml]: Same here.

**Comment by ocefpaf** [comment]
Minor comments and this should be good to go. It would be nice to see what the @conda-forge/fftw and @conda-forge/pyfftw teams think about this recipe and/or join the team here.

**Comment by grlee77** [comment]
@ghisvail, are you interested in being added as a maintainer for either the NFFT or pyNFFT recipes proposed here?  In practice this just gives you commit rights to the recipe feedstocks so that version numbers can be updated as necessary upon release of new versions.  Also, any other feedback you might have on the builds as implemented here would be appreciated.

If you are unfamiliar with [conda-forge](http://conda-forge.github.io), it is a community maintained set of packages with continuous integration for use with the [conda](http://conda.pydata.org) package management system.  A user would install pyNFFT via `conda install -c conda-forge pynfft` (without having to first manually compile FFTW and NFFT as these are also packaged by conda-forge).

**Comment by ghisvail** [comment]
@grlee77 go ahead.

**Comment by grlee77** [comment]
Thanks @ghisvail.  Once this PR is merged feedstocks for each recipe will get automatically generated (at https://github.com/conda-forge/nfft-feedstock, etc.).  You will receive an invite to join conda-forge and once you accept, you will have commit rights to the feedstocks and control over their CIs.

**Comment by ocefpaf** [comment]
@grlee77 This looks good to go! Thanks for adding these BTW I look forward to use them!!

**Pull request merged**